### ISSUE TITLE
Improve Pelias search to search more intelligently worldwide (or find alternative)

### ISSUE THREAD

**Issue by robhawkes** [polygon-city/building-database#8]
Right now the Pelias integration doesn't cope well with worldwide searches. For example, searching for "Taj Mahal" takes you to a random Indian takeaway in Germany. Likewise, searching for "Tower Bridge" doesn't bring up results for Tower Bridge, even if searching with London as the local coordinates.

Whatever search system we use, it should be able to handle the ability to search for a wide-area (eg. London, or Manhattan) as well as local areas (eg. Tower Hamlets, London) and also major landmarks (eg. London Eye, Big Ben, Tower Bridge). In our case we're less interested in searching for a specific address or business name (eg. 123 Example Street, London).

**Comment by bcamper** [comment]
This should be substantially addressed by this pelias PR, which makes geobias (lat-lng or bounding box) optional: https://github.com/pelias/api/pull/66

This demo reflect is -- there are still lots of data issues to work out, especially around duplicate data (working on it... :), but it's much better (for some common cases I tried anyway):
http://rawgit.com/pelias/demo/geobias-optional/index.html

Unfortunately, as mentioned in the PR, there is an elasticsearch bug which prevents this from working in the autocomplete (suggest) endpoint right now. Elasticsearch plans to fix this, but not until v2.0.0, which could be a few months away. A temporary fix could be to treat (abuse :) the full text search endpoint as an autocomplete endpoint, and just call it less frequently on the UI side.

**Comment by bcamper** [comment]
Note a lot of landmarks that are OSM POIs are not returning or ranking highly currently, but this will be addressed.

**Comment by robhawkes** [comment]
It's certainly better! I might be using the demo wrong though as I can't find places like "Berlin" or "Hamburg" for example. It comes up with a tonne of US-based places and other randoms. In fact, I couldn't actually find either being the real cities in Germany.

**Comment by bcamper** [comment]
Yeah, clearly there are several data issues remaining. If you click the button that says "FINE", it changes to "COARSE". The coarse mode is operating on the quattroshapes dataset, which is only administrative areas. These modes will be combined later so that you can search both (and more) data sets.

**Comment by robhawkes** [comment]
Aha, I assumed the little castle icons were landmarks or quite literally castles so didn't click on them with the intention to be taken to a city. Doh!

**Comment by hkrishna** [comment]
@robhawkes I'll update the demo with a legend as to what those castle icons and flags mean. I'll also make a geobias optional switch/button. Oh btw pelias/api#66 has landed so, the API now supports full text search without any lat/lon dependency

**Comment by hkrishna** [comment]
Demo is updated https://github.com/pelias/demo/pull/15 You can play with it here https://rawgit.com/pelias/demo/geobias-toggle/index.html#loc=3,43.64,-5.27&q=turkey&t=coarse&gb=off

**Comment by robhawkes** [comment]
Looking much better, thanks @hkrishna! I think something that would help is putting the country after the place name, to differentiate the duplicates. That's just a UX thing for the demo though, I'll have a proper play with the API updates in Polygon City.

**Comment by robhawkes** [comment]
4d473ef updates the Pelias endpoint and the results are immediately better.

**Comment by robhawkes** [comment]
I've noticed that (obviously) the coarse results mean that the detailed address-level places are excluded. We'll need to look into combining both approaches so you're able to locate a specific address as well as a city-level area.

**Comment by hkrishna** [comment]
Yes, we can locate specific address as well as admin values @robhawkes - just use '/search' instead of '/search/coarse' :smile:

**Comment by robhawkes** [comment]
I'm finally coming back around to adding the Pelias improvements back into Polygon City. I'm still a little unsure about how best to proceed as it seems like the new `/search/coarse` endpoint is for city-level results, whereas it doesn't work at all for address-level results (which is where the `/search` endpoint comes in).

I've also noticed that even then the results aren't really… right. For example, here's a search for London, the city in England:

http://pelias.mapzen.com/search/coarse?input=london&size=5

```
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "id": "36778:_:_:_:london",
        "type": "neighborhood",
        "layer": "neighborhood",
        "name": "London",
        "admin0": "Canada",
        "admin1": "British Columbia",
        "admin2": "Greater Vancouver",
        "text": "London, Greater Vancouver, British Columbia"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -123.160539,
          49.107657
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "id": "15345:_:_:usa:london",
        "type": "neighborhood",
        "layer": "neighborhood",
        "name": "London",
        "alpha3": "USA",
        "admin0": "United States",
        "admin1": "Pennsylvania",
        "admin2": "Mercer",
        "text": "London, Mercer, Pennsylvania"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -80.158997,
          41.139364
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "id": "28180:_:_:can:north_london",
        "type": "neighborhood",
        "layer": "neighborhood",
        "name": "North London",
        "alpha3": "CAN",
        "admin0": "Canada",
        "admin1": "Ontario",
        "admin2": "Middlesex",
        "text": "North London, Middlesex, Ontario"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -81.264719,
          43.00547
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "id": "28179:_:_:can:south_london",
        "type": "neighborhood",
        "layer": "neighborhood",
        "name": "South London",
        "alpha3": "CAN",
        "admin0": "Canada",
        "admin1": "Ontario",
        "admin2": "Middlesex",
        "text": "South London, Middlesex, Ontario"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -81.238403,
          42.966471
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "id": "22666:_:_:gbr:london_wall",
        "type": "neighborhood",
        "layer": "neighborhood",
        "name": "London Wall",
        "alpha3": "GBR",
        "admin0": "United Kingdom",
        "admin1": "England",
        "admin2": "Greater London",
        "text": "London Wall, Greater London"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -0.094519,
          51.517336
        ]
      }
    }
  ],
  "bbox": [
    -123.160539,
    41.139364,
    -0.094519,
    51.517336
  ],
  "date": 1424767766404
}
```

There isn't even a single reference to London the actual city, which surprises me as I'd assume it was in OSM somewhere. The only one referencing London in England is for something called the London Wall, which I'm not even sure people in London would know about =p Even when using the normal `/search` endpoint with a coordinate, the real London isn't there, or if it is it's far down the list. And weirdly, the results for "London, England" don't have a single England location in the top 5.

The other issue is that we also need to support the querying of specific addresses should they exist, or at least road names and local areas. As far as I can see, this requires the `/search` endpoint and so requiring us to make 2 API calls per search. Somehow we need to pull the results from both into a nice UI without cluttering it with the unrelated stuff from each list. This feels a bit clunky.

While I know Pelias is being improved in this regard, my point is more that the results aren't as the user will be expecting for Polygon City (eg. cities ranked above whatever the London Wall is defined as). Perhaps even capital cities ranked above others, etc. And if the UI is cluttered due to us needing to merge the 2 results then it's not very useful.

A perfect API in my mind would have a single call that handled both city-level "London, England" as well as address-level "One Canada Square, London" calls without an origin coordinate. Much like the existing APIs from OSM or Google. For example:

http://open.mapquestapi.com/nominatim/v1/search?q=London&format=json&addressdetails=1&limit=10

```
[
  {
    "place_id": "151142295",
    "licence": "Data © OpenStreetMap contributors, ODbL 1.0. http://www.openstreetmap.org/copyright",
    "osm_type": "relation",
    "osm_id": "65606",
    "boundingbox": [
      "51.2867602",
      "51.6918741",
      "-0.510375",
      "0.3340155"
    ],
    "lat": "51.5073219",
    "lon": "-0.1276474",
    "display_name": "London, Greater London, England, United Kingdom",
    "class": "place",
    "type": "city",
    "importance": 0.9754895765402,
    "icon": "http://open.mapquestapi.com/nominatim/v1/images/mapicons/poi_place_city.p.20.png",
    "address": {
      "city": "London",
      "county": "London",
      "state_district": "Greater London",
      "state": "England",
      "country": "United Kingdom",
      "country_code": "gb"
    }
  },
  {
    "place_id": "4566287",
    "licence": "Data © OpenStreetMap contributors, ODbL 1.0. http://www.openstreetmap.org/copyright",
    "osm_type": "node",
    "osm_id": "485248691",
    "boundingbox": [
      "42.828097",
      "43.148097",
      "-81.4060295",
      "-81.0860295"
    ],
    "lat": "42.988097",
    "lon": "-81.2460295",
    "display_name": "London, Ontario, Canada",
    "class": "place",
    "type": "city",
    "importance": 0.6515723047601,
    "icon": "http://open.mapquestapi.com/nominatim/v1/images/mapicons/poi_place_city.p.20.png",
    "address": {
      "city": "London",
      "state": "Ontario",
      "country": "Canada",
      "country_code": "ca"
    }
  },
  ...
]
```

http://open.mapquestapi.com/nominatim/v1/search?q=One%20Canada%20Square&format=json&addressdetails=1&limit=10

```
[
  {
    "place_id": "41442474",
    "licence": "Data © OpenStreetMap contributors, ODbL 1.0. http://www.openstreetmap.org/copyright",
    "osm_type": "way",
    "osm_id": "5986754",
    "boundingbox": [
      "51.5046739",
      "51.5052126",
      "-0.0199048",
      "-0.0190137"
    ],
    "lat": "51.50493825",
    "lon": "-0.0194592029700664",
    "display_name": "One Canada Square, 1, Canada Square, Isle of Dogs, London Borough of Tower Hamlets, London, Greater London, England, E14 5EW, United Kingdom",
    "class": "place",
    "type": "house",
    "importance": 0.69658137697387,
    "address": {
      "house": "One Canada Square",
      "house_number": "1",
      "road": "Canada Square",
      "suburb": "Isle of Dogs",
      "city": "London",
      "state_district": "Greater London",
      "state": "England",
      "postcode": "E14 5EW",
      "country": "United Kingdom",
      "country_code": "gb"
    }
  }
]
```

The above examples "just work" so I'm inclined to adopt Nominatim until Pelias is at a comparable point.

Any thoughts?

**Comment by missinglink** [comment]
Hi @robhawkes thanks for the feedback, I completely agree with what you are saying and totally 'get' that Pelias is not returning the results you would have expected at this time.

The feedback you provided is really helpful in understanding your specific use-case, so thanks for taking the time to explain in detail. I think your use-case is a fairly common one.

The focus of the Pelias project has shifted this year from 'getting it working' to 'returning quality results' and we are currently working on a testing framework and a definition of 'quality' in order to lay down a clear path to improving search quality over the next few months.

This document outlines some of my thinking on the matter and outlines the 7 types of geocoders organizations commonly require: https://github.com/pelias/pelias/wiki/Search-Quality

To address this issue and similar critique, we are focussing on the use cases outlined in the document above, looking at how other organisations are using the tech and then defining a suite of automated tests which ensure that level of quality is met and maintained.

If the quality is not currently at a level you are happy with then it seems only logical you use nominatum instead, but I would encourage you to file a bug report at https://github.com/pelias/pelias and be involved with the development of the project so that we can all benefit from the results.

**Comment by robhawkes** [comment]
Thanks Peter, it's good to get some more insight into what's going on. I can't wait for the day when Pelias meets and even surpasses the existing solutions, it's just a timing thing that we're having a bit of trouble with it now for Polygon City. To clarify, my comments aren't intended to imply that I expected more today (I know this stuff is complex), I suppose it's more of a wish-list for the future as I'd love to be using Pelias.

I think Polygon City will provide a good use-case for testing the usability of Pelias, as will some things I need to do in ViziCities for that matter. I'm definitely up for using them as test-cases as improvements are made.

It looks like you've got all the things I can think of right now in that Search Quality document (which is awesome btw), so I'm confident that things will improve if that's the yardstick.

I'll have a chat with Brett and the other Peter to see how they feel about this and whether we decide to use something like Nominatim for the time being.

**Comment by robhawkes** [comment]
1e7220e replaces Pelias with Nominatim, at least as a temporary measure until we're ready to switch back.

**Comment by meetar** [comment]
Works for me for now – happy to close this and open a new issue once Pelias updates

**Comment by robhawkes** [comment]
Cool, let's open a new issues once Pelias has progressed.
